Fix error details not being logged in the tracing's span event

filipewl opened the pull request:
#### What is the purpose of this pull request?

Fix how the field "error" of an "error" span event is not displaying the error's details.

|[Before](https://ui.honeycomb.io/vtex/datasets/vtex/result/4gcqfkK2VnT/trace/qrDKt35zRAb?fields%5B%5D=c_name&fields%5B%5D=c_service.name&span=7343fae299df18bb)|[After](https://ui.honeycomb.io/vtex/datasets/vtex/result/wqY64s5khL9/trace/oy4qsTDkA6L?fields%5B%5D=c_name&fields%5B%5D=c_service.name&fields%5B%5D=c_service.name_and_version&span=b3d5f187c40e6b2a)|
|-|-|
|![CleanShot 2023-09-14 at 14 01 28@2x](https://github.com/vtex/node-vtex-api/assets/381395/901ca798-5df3-475b-98a4-17294860bc1e)|![CleanShot 2023-09-19 at 12 24 19](https://github.com/vtex/node-vtex-api/assets/381395/20b3ac26-a3d7-4063-b6c6-81ac65f7a6f6)|

#### What problem is this solving?

It appears this "error" field is categorized as a boolean type, so the [expected error details](https://github.com/vtex/node-error-report/blob/7971b91dcde144b44c3b8c7534bd71a80017b132/src/ErrorReportBase.ts#L108-L120) isn't being used. By renaming this field to "error.details", we expect to have it displayed on Honeycomb.

#### How should this be manually tested?

1. Visit https://filipelimaxtracing--storecomponents.myvtex.com/ where I have yarn-linked this branch to a vtex-linked `pages-graphql` and `render-server` apps.
2. To avoid being sampled, use a browser extension to include the header `jaeger-debug-id=ANY_ID_YOU_WANT_USE_TO_QUERY_HONEYCOMB`. 
3. Navigate through the store to send some traces to Honeycomb.
4. In Honeycomb's Query page, add a WHERE filter to `jaeger-debug-id = ANY_ID_YOU_WANT_USE_TO_QUERY_HONEYCOMB` to find your traces.

[Example](https://ui.honeycomb.io/vtex/datasets/vtex/result/wqY64s5khL9/trace/oy4qsTDkA6L?fields%5B%5D=c_name&fields%5B%5D=c_service.name&fields%5B%5D=c_service.name_and_version&span=b3d5f187c40e6b2a) of a trace with error span. Check the "Span events" tab on the side bar and expand the "error" event.

#### Types of changes

* [x] Bug fix (a non-breaking change which fixes an issue)
* [ ] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

filipewl commented on src/tracing/errorReporting/ErrorReport.ts:
```diff
@@ -61,7 +61,7 @@ export class ErrorReport extends ErrorReportBase {
     }
 
     const serializableError = this.toObject()
```
`toObject` should return some [JS object with error details](https://github.com/vtex/node-error-report/blob/7971b91dcde144b44c3b8c7534bd71a80017b132/src/ErrorReportBase.ts#L108-L119).

filipewl commented on src/tracing/errorReporting/ErrorReport.ts:
```diff
@@ -61,7 +61,7 @@ export class ErrorReport extends ErrorReportBase {
 
     const serializableError = this.toObject()
-    span.log({ event: 'error', ...indexedLogs, error: serializableError })
+    span.log({ event: 'error', ...indexedLogs, 'error.details': serializableError })
```
Instead of the JS object mentioned above, we're always getting a boolean for the field `error`. I think by using a different key, that doesn't conflict with a general "error" name, should fix it.

mairatma commented on src/tracing/errorReporting/ErrorReport.ts:
```diff
@@ -61,7 +61,14 @@ export class ErrorReport extends ErrorReportBase {
+      event: 'error',
+      ...indexedLogs,
+      [ErrorReportLogFields.ERROR_MESSAGE]: serializableError.message,
+      [ErrorReportLogFields.ERROR_METADATA]: serializableError.metadata,
```
Do we need to manually serialize this as well? It returns an object from its type definition.

filipewl replied:
Done [here](https://github.com/vtex/node-vtex-api/pull/543/commits/eadfe42ce1b461556c1c13c46a1214b8120fa0d0).

filipewl commented:
> @filipewl it might be worth talking to someone from the Observability team to see how to test this. Isn't there a way to force a trace to be created for a request to linked code?

I figured out this with @arturpimentel. There's a header `jaeger-debug-id` that you can include in the requests so they won't be sampled and that can be later used on Honeycomb to find your traces. I've updated the PR's description to explain how to test and added a link to a tracing example.

mairatma reviewed (approved):
Awesome, thanks! Lgtm

mairatma commented on src/tracing/errorReporting/ErrorReport.ts:
```diff
@@ -65,7 +65,8 @@ export class ErrorReport extends ErrorReportBase {
       ...indexedLogs,
       [ErrorReportLogFields.ERROR_MESSAGE]: serializableError.message,
-      [ErrorReportLogFields.ERROR_METADATA]: serializableError.metadata,
+      [ErrorReportLogFields.ERROR_METADATA_METRICS_INSTANTIATION_TIME]: serializableError.metadata.reportCount,
```
@filipewl is there any chance that `metadata` may be null or undefined? Just making sure we won't throw here by mistake.

filipewl replied:
Type-wise, it doesn't look like it can, but I'm afraid of runtime data... I added some optional chaining [here](https://github.com/vtex/node-vtex-api/pull/543/commits/dc37cf0e64500ca898a408cc1314742527fd88dd) to be safe.

mairatma replied:
Also afraid, thanks!
